File: behavioral/interpreter.php
<?php
interface Expression {
    public function interpret(array $context): int;
}

class Number implements Expression {
    public function __construct(private int $number){}
    public function interpret(array $context): int { return $this->number; }
}

class Plus implements Expression {
    public function __construct(private Expression $left, private Expression $right){}
    public function interpret(array $context): int {
        return $this->left->interpret($context) + $this->right->interpret($context);
    }
}

// (5 + 10)
$expr = new Plus(new Number(5), new Number(10));
echo $expr->interpret([]); // 15

File: behavioral/interpreter.go
package main

import "fmt"

type Expression interface{ Interpret() int }

type Number struct{ value int }

func (n Number) Interpret() int { return n.value }

type Plus struct{ left, right Expression }

func (p Plus) Interpret() int { return p.left.Interpret() + p.right.Interpret() }

func main() {
	expr := Plus{Number{5}, Number{10}}
	fmt.Println(expr.Interpret()) // 15
}
